src/compiler/checker.ts

```diff
@@ -13692,23 +13705,56 @@ namespace ts {
                 const numberIndexInfo = hasComputedNumberProperty && !isJSObjectLiteral ? getObjectLiteralIndexInfo(node.properties, offset, propertiesArray, IndexKind.Number) : undefined;
-                const result = createAnonymousType(node.symbol, propertiesTable, emptyArray, emptyArray, stringIndexInfo, numberIndexInfo);
-                const freshObjectLiteralFlag = compilerOptions.suppressExcessPropertyErrors ? 0 : TypeFlags.FreshLiteral;
-                result.flags |= TypeFlags.ContainsObjectLiteral | freshObjectLiteralFlag | (typeFlags & TypeFlags.PropagatingFlags);
-                result.objectFlags |= ObjectFlags.ObjectLiteral;
+                // We used a cached object literal type if the object literal is not the inferred type of a variable
+                // (otherwise Go To Definition on the variable's members wouldn't work properly), has no methods or
+                // accessors (we may cause circularities when obtaining the types of those members), has no index
+                // signatures, is not in a destructuring pattern, doesn't have more than 10 properties, and doesn't
+                // have property names containing colons (which we use as a separator in the cache key).
+                if (!isInferredTypeOfVariable(node) && !hasMethodsOrAccessors && !stringIndexInfo && !numberIndexInfo && !inDestructuringPattern &&
+                    propertiesArray.length <= 10 && every(propertiesArray, propertyNameIsValidKey)) {
+                    return getObjectLiteralType(typeFlags, node.symbol, propertiesTable, propertiesArray);
+                }
+                const result = makeObjectLiteralType(typeFlags, node.symbol, propertiesTable, stringIndexInfo, numberIndexInfo);
                 if (patternWithComputedProperties) {
                     result.objectFlags |= ObjectFlags.ObjectLiteralPatternWithComputedProperties;
                 }
                 if (inDestructuringPattern) {
                     result.pattern = node;
                 }
-                if (!(result.flags & TypeFlags.Nullable)) {
-                    propagatedFlags |= (result.flags & TypeFlags.PropagatingFlags);
-                }
                 return result;
             }
         }
 
+        function isInferredTypeOfVariable(node: Node) {
+            while (node.parent.kind === SyntaxKind.PropertyAssignment) {
+                node = node.parent.parent;
+            }
+            const parent = node.parent;
+            return isVariableLike(parent) && !parent.type && parent.initializer === node;
+        }
+
+        function propertyNameIsValidKey(symbol: Symbol) {
+            return (<string>symbol.escapedName).indexOf(":") < 0;
+        }
+
+        function getObjectLiteralType(flags: TypeFlags, symbol: Symbol, members: SymbolTable, properties: Symbol[]): ResolvedType {
+            const key = map(properties, p => p.escapedName + (p.flags & SymbolFlags.Optional ? ":?" : ":") + getTypeOfSymbol(p).id).join(",");

```

Comment:
calling `getTypeOfSymbol` here is a lot more eager than before. Probably not a problem?

Reply:
Not a problem, any symbol we touch here already had its `type` property explicitly initialized by the code that calls this function.